aiida/manage/external/pgsu.py

```diff
@@ -0,0 +1,348 @@
+        :param kwargs: will be forwarded to _execute_... function
+        """
+        # Use self.dbinfo as default kwargs, update with provided kwargs
+        kw_copy = self.dbinfo.copy()
+        kw_copy.update(kwargs)
+
+        if self.connection_mode == PostgresConnectionMode.PSYCOPG:  # pylint: disable=no-else-return
+            return _execute_psyco(command, **kw_copy)
+        elif self.connection_mode == PostgresConnectionMode.PSQL:
+            return _execute_psql(command, **kw_copy)
+
+        raise ValueError('Could not connect to postgres.')
+
+    def set_setup_fail_callback(self, callback):
+        """
+        Set a callback to be called when setup cannot be determined automatically
+
+        :param callback: a callable with signature ``callback(interactive, dbinfo)``
+          that returns a ``dbinfo`` dictionary.
+        """
+        self.setup_fail_callback = callback
+
+    def determine_setup(self):
+        """Determine how to connect as the postgres superuser.
+
+        Depending on how postgres is set up, psycopg2 can be used to create dbs and db users,
+        otherwise a subprocess has to be used that executes psql as an os user with appropriate permissions.
+
+        Note: We aim to connect as a superuser (typically 'postgres') with privileges to manipulate (create/drop)
+          databases and database users.
+
+        :returns success: True, if connection could be established.
+        :rtype success: bool
+        """
+        # find out if we run as a postgres superuser or can connect as postgres
+        # This will work on OSX in some setups but not in the default Debian one
+        dbinfo = self.dbinfo.copy()
+
+        for pg_user in set([dbinfo.get('user'), None]):
+            dbinfo['user'] = pg_user
+            if _try_connect_psycopg(**dbinfo):
+                self.dbinfo = dbinfo
+                self.connection_mode = PostgresConnectionMode.PSYCOPG
+                return True
+
+        # This will work for the default Debian postgres setup, assuming that sudo is available to the user
+        # Check if the user can find the sudo command
+        if _sudo_exists():
+            if _try_subcmd(interactive=self.interactive, quiet=self.quiet, **dbinfo):
+                self.dbinfo = dbinfo
+                self.connection_mode = PostgresConnectionMode.PSQL
+                return True
+        elif not self.quiet:
+            click.echo('Warning: Could not find `sudo` for connecting to the database.')
+
+        self.setup_fail_counter += 1
+        return self._no_setup_detected()
+
+    def _no_setup_detected(self):
+        """Print a warning message and calls the failed setup callback
+
+        :returns: False, if no successful try.
+        """
+        message = '\n'.join([
+            'Warning: Unable to autodetect postgres setup - do you know how to access it?',
+        ])
+
+        if not self.quiet:
+            click.echo(message)
```

Comment:
Maybe use `echo_warning` instead here

Reply:
As mentioned at the top of the file, the idea is to move this out of aiida-core eventually so it can be easily tested.

Reply:
Ah of course, missed that, thanks